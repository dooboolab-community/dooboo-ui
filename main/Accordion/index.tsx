import {Animated, StyleProp, ViewStyle} from 'react-native';
import React, {FC, useRef} from 'react';

import AccordionItem from './AccordionItem';
import styled from '@emotion/native';

const Container = styled.View`
  flex-direction: column;
  align-items: center;
`;

export interface Datum {
  title: string;
  bodies: string[];
}

export type Styles = {
  titleContainer?: StyleProp<ViewStyle>;
  bodyContainer?: StyleProp<ViewStyle>;
};

interface Props {
  style?: ViewStyle;
  styles?: Styles;
  data: Datum[];
  shouldAnimate?: boolean;
  collapseOnStart?: boolean;
  animDuration?: number;
  activeOpacity?: number;
  toggleElement?: React.ReactElement;
  renderTitle?: (item: string) => React.ReactElement;
  renderBody?: (item: string) => React.ReactElement;
}

const Accordion: FC<Props> = (props) => {
  const {
    style,
    styles,
    data,
    collapseOnStart,
    shouldAnimate,
    animDuration,
    activeOpacity,
    toggleElement,
    renderTitle,
    renderBody,
  } = props;

  const dropDownAnimValueList = useRef(
    data.map(() => new Animated.Value(0)),
  ).current;

  return (
    <Container style={style}>
      {data.map((datum, titleKey) => {
        return (
          <AccordionItem
            key={titleKey}
            testID={`${titleKey}`}
            styles={styles}
            datum={datum}
            collapseOnStart={collapseOnStart}
            shouldAnimate={shouldAnimate}
            animDuration={animDuration}
            activeOpacity={activeOpacity}
            dropDownAnimValueList={dropDownAnimValueList[titleKey]}
            sumOfPrecedingTranslateY={dropDownAnimValueList
              .filter((item, idx) => idx < titleKey)
              .map((value) => ({translateY: value}))}
            toggleElement={toggleElement}
            renderTitle={renderTitle}
            renderBody={renderBody}
          />
        );
      })}
    </Container>
  );
};

export {Accordion};
