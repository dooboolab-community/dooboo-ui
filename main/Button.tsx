import {
  ActivityIndicator,
  LayoutRectangle,
  Platform,
  TouchableOpacity,
} from 'react-native';
import {DoobooTheme, light, withTheme} from './theme';
import React, {useRef, useState} from 'react';
import type {
  StyleProp,
  TextProps,
  TextStyle,
  TouchableOpacityProps,
  ViewStyle,
} from 'react-native';

import type {FC} from 'react';
import styled from '@emotion/native';
import {useHover} from 'react-native-web-hooks';

type Styles = {
  container?: StyleProp<ViewStyle>;
  text?: StyleProp<TextStyle>;
  disabledButton?: StyleProp<ViewStyle>;
  disabledText?: StyleProp<TextStyle>;
  hovered?: StyleProp<ViewStyle>;
};

const Container = styled.View`
  align-self: stretch;
  padding: 8px 12px;
  background-color: ${({theme}) => theme.primary || light.primary};

  flex-direction: row;
  align-items: center;
  justify-content: center;
`;

const Text = styled.Text`
  font-size: 14px;
  color: white;
`;

export interface ButtonProps {
  testID?: string;
  indicatorColor?: string;
  loading?: boolean;
  disabled?: boolean;
  style?: StyleProp<ViewStyle>;
  styles?: Styles;
  leftElement?: React.ReactElement;
  rightElement?: React.ReactElement;
  activeOpacity?: TouchableOpacityProps['activeOpacity'];
  text?: string;
  onPress?: TouchableOpacityProps['onPress'];
  touchableOpacityProps?: Partial<TouchableOpacityProps>;
  textProps?: Partial<TextProps>;
}

const StyledButton: FC<ButtonProps & {theme: DoobooTheme}> = ({
  testID,
  theme,
  disabled,
  loading,
  style,
  styles,
  indicatorColor = theme.disabled,
  leftElement,
  rightElement,
  activeOpacity = 0.7,
  text,
  onPress,
  touchableOpacityProps,
  textProps,
}) => {
  const ref = useRef<TouchableOpacity>(null);
  const hovered = useHover(ref);
  const [layout, setLayout] = useState<LayoutRectangle>();

  const compositeStyles: Styles = {
    disabledButton: {
      backgroundColor: theme.primary,
      borderColor: theme.primary,
    },
    disabledText: {
      color: theme.primary,
    },
    hovered: {
      borderColor: theme.primary,
      shadowColor: 'black',
      shadowOffset: {
        width: 0,
        height: 4,
      },
      shadowOpacity: 0.24,
      shadowRadius: 16,
      elevation: 10,
    },
    ...styles,
  };

  return (
    <TouchableOpacity
      testID={testID}
      ref={Platform.select({
        web: ref,
        default: undefined,
      })}
      activeOpacity={activeOpacity}
      onPress={onPress}
      delayPressIn={50}
      disabled={disabled}
      style={style}
      {...touchableOpacityProps}>
      {loading ? (
        <Container
          testID="loading-view"
          style={[
            compositeStyles.container,
            {
              width: layout?.width,
              height: layout?.height,
            },
            hovered && !disabled && compositeStyles.hovered,
            disabled && compositeStyles.disabledButton,
          ]}>
          <ActivityIndicator size="small" color={indicatorColor} />
        </Container>
      ) : (
        <Container
          testID="button-view"
          style={[
            compositeStyles.container,
            hovered && !disabled && compositeStyles.hovered,
            disabled && compositeStyles.disabledButton,
          ]}
          onLayout={(e) => setLayout(e.nativeEvent.layout)}>
          {leftElement}
          <Text
            style={[
              compositeStyles.text,
              disabled && compositeStyles.disabledText,
            ]}
            {...textProps}>
            {text}
          </Text>
          {rightElement}
        </Container>
      )}
    </TouchableOpacity>
  );
};

StyledButton.defaultProps = {theme: light};

export const Button = withTheme(StyledButton);
