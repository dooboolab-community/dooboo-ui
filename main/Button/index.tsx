import {
  ActivityIndicator,
  LayoutRectangle,
  Platform,
  TouchableOpacity,
} from 'react-native';
import {DoobooTheme, light, useTheme, withTheme} from '../theme';
import React, {useRef, useState} from 'react';
import type {
  StyleProp,
  TextProps,
  TextStyle,
  TouchableOpacityProps,
  ViewStyle,
} from 'react-native';

import type {FC} from 'react';
import {TypographyInverted} from '../Typography';
import styled from '@emotion/native';
import {useHover} from 'react-native-web-hooks';

type Styles = {
  container?: StyleProp<ViewStyle>;
  text?: StyleProp<TextStyle>;
  disabledButton?: StyleProp<ViewStyle>;
  disabledText?: StyleProp<TextStyle>;
  hovered?: StyleProp<ViewStyle>;
};

type ButtonType = 'primary' | 'secondary' | 'danger' | 'warning' | 'info';
type ButtonSize = 'small' | 'medium' | 'large';

const ButtonContainer = styled.View<{
  type: ButtonType;
  size?: ButtonSize;
  outlined?: boolean;
}>`
  align-self: stretch;
  padding: ${({size}) =>
    size === 'large'
      ? '14px 24px'
      : size === 'small'
      ? '4px 12px'
      : '10px 20px'};
  border-radius: ${({size}) => (size === 'large' ? '24px' : '20px')};
  border-width: ${({outlined}) => (outlined ? '1px' : 0)};
  background-color: ${({theme, type, outlined}) =>
    outlined
      ? theme.background
      : type === 'info'
      ? theme.info
      : type === 'secondary'
      ? theme.secondary
      : type === 'danger'
      ? theme.danger
      : type === 'warning'
      ? theme.warning
      : theme.primary};
  border-color: ${({theme, type}) =>
    type === 'info'
      ? theme.info
      : type === 'secondary'
      ? theme.secondary
      : type === 'danger'
      ? theme.danger
      : type === 'warning'
      ? theme.warning
      : theme.primary};

  flex-direction: row;
  align-items: center;
  justify-content: center;
`;

export interface ButtonProps {
  testID?: string;
  indicatorColor?: string;
  loading?: boolean;
  disabled?: boolean;
  outlined?: boolean;
  style?: StyleProp<ViewStyle>;
  styles?: Styles;
  leftElement?: React.ReactElement;
  rightElement?: React.ReactElement;
  activeOpacity?: TouchableOpacityProps['activeOpacity'];
  text?: string;
  onPress?: TouchableOpacityProps['onPress'];
  touchableOpacityProps?: Partial<TouchableOpacityProps>;
  textProps?: Partial<TextProps>;
  type?: ButtonType;
  size?: ButtonSize;
}

const StyledButton: FC<ButtonProps & {theme: DoobooTheme}> = ({
  testID,
  theme,
  disabled,
  loading,
  style,
  styles,
  indicatorColor = theme.disabled,
  leftElement,
  rightElement,
  activeOpacity = 0.6,
  text,
  onPress,
  touchableOpacityProps,
  textProps,
  type = 'primary',
  size,
  outlined = false,
}) => {
  const ref = useRef<TouchableOpacity>(null);
  const hovered = useHover(ref);
  const [layout, setLayout] = useState<LayoutRectangle>();
  const {themeType} = useTheme();

  const mainColor =
    type === 'info'
      ? theme.info
      : type === 'secondary'
      ? theme.secondary
      : type === 'danger'
      ? theme.danger
      : type === 'warning'
      ? theme.warning
      : theme.primary;

  const compositeStyles: Styles = {
    disabledButton: {
      backgroundColor: theme.disabled,
      borderColor: theme.disabled,
    },
    disabledText: {
      color: theme.disabled,
    },
    hovered: {
      shadowColor: 'black',
      shadowOffset: {
        width: 0,
        height: 4,
      },
      shadowOpacity: 0.24,
      shadowRadius: 16,
      elevation: 10,
    },
    ...styles,
  };

  return (
    <TouchableOpacity
      testID={testID}
      ref={Platform.select({
        web: ref,
        default: undefined,
      })}
      activeOpacity={activeOpacity}
      onPress={onPress}
      delayPressIn={50}
      disabled={disabled}
      style={style}
      {...touchableOpacityProps}>
      {loading ? (
        <ButtonContainer
          testID="loading-view"
          style={[
            compositeStyles.container,
            {
              width: layout?.width,
              height: layout?.height,
            },
            hovered && !disabled && compositeStyles.hovered,
            disabled && compositeStyles.disabledButton,
          ]}
          type={type}
          size={size}
          outlined={outlined}>
          <ActivityIndicator size="small" color={indicatorColor} />
        </ButtonContainer>
      ) : (
        <ButtonContainer
          testID="button-container"
          style={[
            compositeStyles.container,
            hovered && !disabled && compositeStyles.hovered,
            disabled && compositeStyles.disabledButton,
          ]}
          onLayout={(e) => setLayout(e.nativeEvent.layout)}
          type={type}
          size={size}
          outlined={outlined}>
          {leftElement}
          <TypographyInverted.Heading3
            style={[
              {
                color:
                  themeType === 'dark' && !outlined
                    ? theme.textContrast
                    : outlined && type === 'warning'
                    ? theme.text
                    : outlined
                    ? mainColor
                    : type !== 'primary'
                    ? theme.text
                    : theme.textContrast,
              },
              compositeStyles.text,
              disabled && compositeStyles.disabledText,
            ]}
            {...textProps}>
            {text}
          </TypographyInverted.Heading3>
          {rightElement}
        </ButtonContainer>
      )}
    </TouchableOpacity>
  );
};

StyledButton.defaultProps = {theme: light};

export const Button = withTheme(StyledButton);
