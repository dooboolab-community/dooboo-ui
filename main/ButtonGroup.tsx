import {DoobooTheme, light, useTheme, withTheme} from './theme';
import {
  StyleProp,
  StyleSheet,
  Text,
  TextStyle,
  TouchableOpacity,
  View,
  ViewStyle,
} from 'react-native';

interface Styles {
  container: StyleProp<ViewStyle>;
  button: ViewStyle;
  selectedButton: ViewStyle;
  text: StyleProp<TextStyle>;
  selectedText: StyleProp<TextStyle>;
}

interface Props<T> {
  testID?: string;
  theme: DoobooTheme;
  borderRadius?: number;
  borderWidth?: number;
  style?: StyleProp<ViewStyle>;
  styles?: Styles;
  data: T[];
  color?: string;
  onPress?: (i: number) => void;
  selected?: number;
}

function StyledButtonGroup<T>(props: Props<T>): React.ReactElement {
  const {theme} = useTheme();

  const {
    borderRadius = 0,
    selected = 0,
    borderWidth = 1,
    color = theme.text,
    testID,
    style,
    data,
    onPress,
    styles,
  } = props;

  return (
    <View
      testID={testID}
      style={StyleSheet.flatten([
        {borderColor: color},
        styles?.container,
        style,
      ])}>
      {data.map((text, i) => {
        return (
          <TouchableOpacity
            key={i}
            testID={`CHILD_${i}`}
            activeOpacity={0.85}
            style={{flex: 1}}
            onPress={(): void => {
              if (onPress) onPress(i);
            }}>
            <View
              style={StyleSheet.flatten([
                selected === i
                  ? {...styles?.selectedButton, backgroundColor: color}
                  : {...styles?.button, borderColor: color},
                i === 0
                  ? {
                      borderLeftWidth: borderWidth,
                      borderTopWidth: borderWidth,
                      borderBottomWidth: borderWidth,
                      borderTopLeftRadius: borderRadius,
                      borderBottomLeftRadius: borderRadius,
                    }
                  : i === data.length - 1
                  ? {
                      borderRightWidth: borderWidth,
                      borderLeftWidth: borderWidth,
                      borderTopWidth: borderWidth,
                      borderBottomWidth: borderWidth,
                      borderBottomRightRadius: borderRadius,
                      borderTopRightRadius: borderRadius,
                    }
                  : {
                      borderLeftWidth: borderWidth,
                      borderTopWidth: borderWidth,
                      borderBottomWidth: borderWidth,
                    },
                {borderColor: color},
              ])}>
              <Text
                style={
                  selected === i
                    ? [styles?.selectedText, {color: theme.textContrast}]
                    : [styles?.text, {color: theme.text}]
                }>
                {text}
              </Text>
            </View>
          </TouchableOpacity>
        );
      })}
    </View>
  );
}

StyledButtonGroup.defaultProps = {
  theme: light,
  styles: {
    container: {
      backgroundColor: 'transparent',
      flexDirection: 'row',
      alignItems: 'center',
      justifyContent: 'center',
      alignSelf: 'stretch',
      minHeight: 40,
      marginTop: 24,
    },
    button: {
      alignSelf: 'stretch',
      minHeight: 40,

      justifyContent: 'center',
      alignItems: 'center',
    },
    selectedButton: {
      alignSelf: 'stretch',
      minHeight: 40,

      justifyContent: 'center',
      alignItems: 'center',
    },
    text: {
      fontSize: 14,
      textAlign: 'center',
      alignSelf: 'center',
    },
    selectedText: {
      fontSize: 14,
      textAlign: 'center',
      alignSelf: 'center',
    },
  },
  data: ['option 1', 'option 2'],
};

export const ButtonGroup = withTheme<any>(StyledButtonGroup);
