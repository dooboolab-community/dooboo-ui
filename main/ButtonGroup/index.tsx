import React from 'react';
import {
  StyleProp,
  StyleSheet,
  Text,
  TextStyle,
  TouchableOpacity,
  View,
  ViewStyle,
} from 'react-native';

import {useTheme} from '@dooboo-ui/theme';

interface Styles {
  container: StyleProp<ViewStyle>;
  button: ViewStyle;
  selectedButton: ViewStyle;
  text: StyleProp<TextStyle>;
  selectedText: StyleProp<TextStyle>;
}

interface Props<T> {
  testID?: string;
  borderRadius?: number;
  borderWidth?: number;
  style?: StyleProp<ViewStyle>;
  styles?: Styles;
  data: T[];
  color?: string;
  onPress?: (i: number) => void;
  selectedIndex?: number;
}

export function ButtonGroup<T>(props: Props<T>): React.ReactElement {
  const {theme} = useTheme();

  const {
    borderRadius = 0,
    selectedIndex = 0,
    borderWidth = 1,
    color = theme.text,
    testID,
    style,
    data,
    onPress,
    styles,
  } = props;

  const borderWidthAndRadius = (index: number): object => {
    const fullWidthAndRadius = {
      borderLeftWidth: borderWidth,
      borderRightWidth: borderWidth,
      borderTopWidth: borderWidth,
      borderBottomWidth: borderWidth,

      borderTopLeftRadius: borderRadius,
      borderBottomLeftRadius: borderRadius,
      borderTopRightRadius: borderRadius,
      borderBottomRightRadius: borderRadius,
    };

    const isFirst = index === 0;
    const isLast = index === data.length - 1;

    const borderForFirstElement = {
      ...fullWidthAndRadius,
      borderTopRightRadius: undefined,
      borderBottomRightRadius: undefined,
    };

    const borderForLastElement = {
      ...fullWidthAndRadius,
      borderTopLeftRadius: undefined,
      borderBottomLeftRadius: undefined,
    };

    const borderForMiddleElement = {
      borderRightWidth: borderWidth,
      borderTopWidth: borderWidth,
      borderBottomWidth: borderWidth,
    };

    if (data.length === 1) return fullWidthAndRadius;

    if (isFirst) return borderForFirstElement;

    if (isLast) {
      if (data.length === 2)
        return {
          ...borderForLastElement,
          borderLeftWidth: undefined,
        };

      return borderForLastElement;
    }

    return borderForMiddleElement;
  };

  return (
    <View
      testID={testID}
      style={StyleSheet.flatten([
        {borderColor: color},
        styles?.container,
        style,
      ])}
    >
      {data.map((text, i) => {
        return (
          <TouchableOpacity
            key={i}
            activeOpacity={0.85}
            style={{flex: 1}}
            onPress={(): void => {
              if (onPress) onPress(i);
            }}
          >
            <View
              testID={`CHILD_${i}`}
              style={StyleSheet.flatten([
                selectedIndex === i
                  ? {...styles?.selectedButton, backgroundColor: color}
                  : {...styles?.button, borderColor: color},
                borderWidthAndRadius(i),
                {borderColor: color},
              ])}
            >
              <Text
                style={
                  selectedIndex === i
                    ? [styles?.selectedText, {color: theme.textContrast}]
                    : [styles?.text, {color: theme.text}]
                }
              >
                {text}
              </Text>
            </View>
          </TouchableOpacity>
        );
      })}
    </View>
  );
}

ButtonGroup.defaultProps = {
  styles: {
    container: {
      backgroundColor: 'transparent',
      flexDirection: 'row',
      alignItems: 'center',
      justifyContent: 'center',
      alignSelf: 'stretch',
      minHeight: 40,
      marginTop: 24,
    },
    button: {
      alignSelf: 'stretch',
      minHeight: 40,

      justifyContent: 'center',
      alignItems: 'center',
    },
    selectedButton: {
      alignSelf: 'stretch',
      minHeight: 40,

      justifyContent: 'center',
      alignItems: 'center',
    },
    text: {
      fontSize: 14,
      textAlign: 'center',
      alignSelf: 'center',
    },
    selectedText: {
      fontSize: 14,
      textAlign: 'center',
      alignSelf: 'center',
    },
  },
  data: ['option 1', 'option 2'],
};
