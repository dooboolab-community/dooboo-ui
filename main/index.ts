export * from './uis/Typography';
export * from './uis/RadioGroup';
export * from './modals/Snackbar';
export * from './uis/Icon';
export * from './uis/Hr';
export * from './uis/Button';
export * from './uis/ButtonGroup';
export * from './uis/IconButton';
export * from './uis/Checkbox';
export * from './uis/EditText';
export * from './uis/NetworkImage';
export * from './uis/ProgressBar';
export * from './uis/ProgressCircle';
export * from './uis/LoadingIndicator';
export * from './uis/Accordion';
export * from './uis/StatusbarBrightness';
export * from './uis/SwitchToggle';
export * from './uis/Fabs';
export * from './uis/Rating';
export * from './uis/CustomPressable';
export * from './providers';
