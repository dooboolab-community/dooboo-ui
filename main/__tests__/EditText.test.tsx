import React, {ReactElement} from 'react';
import {RenderAPI, act, fireEvent, render} from '@testing-library/react-native';

import {EditText} from '../../main';
import type {EditTextProps} from '../../main/EditText';
import RNWebHooks from 'react-native-web-hooks';
import {View} from 'react-native';
import {createComponent} from '../../test/testUtils';

jest.mock('react-native-web-hooks', () => ({
  useHover: () => false,
}));

let testingLib: RenderAPI;

const component = (editProps?: EditTextProps): ReactElement =>
  createComponent(<EditText {...editProps} />);

describe('[EditText]', () => {
  jest.spyOn(console, 'error').mockImplementation(() => {});

  describe('interactions', () => {
    beforeAll(() => {
      testingLib = render(
        component({
          autoCapitalize: 'words',
        }),
      );
    });

    describe('hovered', () => {
      beforeAll(() => {
        jest.spyOn(RNWebHooks, 'useHover').mockImplementation(() => true);
      });

      it('should render styles', async () => {
        testingLib = render(
          component({
            styles: {
              hovered: {
                backgroundColor: 'green',
              },
            },
          }),
        );

        const container = testingLib.getByTestId('container-id');

        const containerChildViewCustomStyle =
          // @ts-ignore
          container.findAllByType(View)[0].instance.props.style[1][1];

        expect(containerChildViewCustomStyle).toEqual({
          backgroundColor: 'green',
        });
      });

      describe('labeText', () => {
        it('should render text', async () => {
          testingLib = render(
            component({
              labelText: 'label text',
            }),
          );

          const label = testingLib.getByText('label text');

          expect(label).toBeTruthy();
        });

        it('should render style', async () => {
          testingLib = render(
            component({
              labelText: 'label text',
              styles: {
                labelText: {
                  color: 'green',
                },
              },
            }),
          );

          const label = testingLib.getByText('label text');
          const labelTextStyle = label.props.style[0][1];

          expect(labelTextStyle).toEqual({color: 'green'});
        });

        describe('unhovered', () => {
          beforeAll(() => {
            jest.spyOn(RNWebHooks, 'useHover').mockImplementation(() => false);
          });

          it('should contain `focusColor` when focused', async () => {
            testingLib = render(
              component({
                testID: 'INPUT_TEST',
                labelText: 'label text',
                styles: {
                  labelText: {
                    color: 'green',
                  },
                },
                disableColor: '#666',
                focusColor: 'pink',
              }),
            );

            const input = testingLib.getByTestId('INPUT_TEST');

            act(() => {
              input.props.onFocus();
            });

            const label = testingLib.getByText('label text');

            const unhoveredTextStyle = label.props.style[2];

            expect(unhoveredTextStyle).toEqual({color: 'pink'});
          });

          it('should contain `errorColor` when `errorText is provided', async () => {
            testingLib = render(
              component({
                testID: 'INPUT_TEST',
                labelText: 'label text',
                errorText: 'error text',
                styles: {
                  labelText: {
                    color: 'green',
                  },
                },
                errorColor: 'orange',
              }),
            );

            const input = testingLib.getByTestId('INPUT_TEST');

            act(() => {
              input.props.onFocus();
            });

            const label = testingLib.getByText('label text');
            const error = testingLib.getByText('error text');

            const unhoveredTextStyle = label.props.style[2];

            expect(error).toBeTruthy();
            expect(unhoveredTextStyle).toEqual({color: 'orange'});
          });
        });
      });

      it('should trigger `onFocus`', async () => {
        testingLib = render(
          component({
            testID: 'INPUT_TEST',
            onFocus: jest.fn(),
          }),
        );

        const input = testingLib.getByTestId('INPUT_TEST');

        expect(input).toBeTruthy();

        fireEvent(input, 'focus');
      });

      describe('onBlur (focused === false)', () => {
        it('should trigger blur without errorText', async () => {
          testingLib = render(
            component({
              onBlur: () => {},
              testID: 'INPUT_TEST',
            }),
          );

          const input = testingLib.getByTestId('INPUT_TEST');

          expect(input).toBeTruthy();

          fireEvent(input, 'blur');
        });

        it('should trigger blur with errorText', async () => {
          testingLib = render(
            component({
              testID: 'INPUT_TEST',
              errorText: 'error text',
            }),
          );

          const input = testingLib.getByTestId('INPUT_TEST');

          expect(input).toBeTruthy();

          fireEvent(input, 'blur');
        });
      });
    });

    describe('Type: [column] - default', () => {
      it('renders without crashing', () => {
        testingLib = render(
          component({
            type: 'column',
          }),
        );

        const json = testingLib.toJSON();

        expect(json).toMatchSnapshot();
        expect(json).toBeTruthy();
      });

      it('renders custom styles without crashing', () => {
        testingLib = render(
          component({
            type: 'column',
            styles: {
              container: {},
              errorText: {},
              hovered: {},
              input: {},
              labelText: {},
              labelTextHovered: {},
              counter: {},
            },
          }),
        );

        const json = testingLib.toJSON();

        expect(json).toMatchSnapshot();
        expect(json).toBeTruthy();
      });

      it('should render `maxLength` counter', () => {
        testingLib = render(component());

        const json = testingLib.toJSON();

        expect(json).toBeTruthy();

        testingLib = render(
          component({
            type: 'column',
            textInputProps: {
              maxLength: 10,
            },
          }),
        );

        expect(json).toMatchSnapshot();
        expect(json).toBeTruthy();
      });

      it('should render `maxLength` counter with error style', () => {
        testingLib = render(component());

        const json = testingLib.toJSON();

        expect(json).toBeTruthy();

        testingLib = render(
          component({
            type: 'column',
            textInputProps: {
              maxLength: 10,
            },
            value: '0123456789',
          }),
        );

        expect(json).toMatchSnapshot();
        expect(json).toBeTruthy();
      });
    });

    describe('Type: [boxed]', () => {
      it('should render without crashing', () => {
        testingLib = render(component());

        const json = testingLib.toJSON();

        expect(json).toBeTruthy();

        testingLib = render(
          component({
            type: 'boxed',
          }),
        );

        expect(json).toMatchSnapshot();
        expect(json).toBeTruthy();
      });

      it('should render `errorText`', () => {
        testingLib = render(component());

        const json = testingLib.toJSON();

        expect(json).toBeTruthy();

        testingLib = render(
          component({
            type: 'boxed',
            errorText: 'error',
          }),
        );

        expect(json).toMatchSnapshot();
        expect(json).toBeTruthy();
      });

      it('should render `!editable` status', () => {
        testingLib = render(component());

        const json = testingLib.toJSON();

        expect(json).toBeTruthy();

        testingLib = render(
          component({
            type: 'boxed',
            editable: false,
          }),
        );

        expect(json).toMatchSnapshot();
        expect(json).toBeTruthy();
      });

      it('should render `maxLength` counter', () => {
        testingLib = render(component());

        const json = testingLib.toJSON();

        expect(json).toBeTruthy();

        testingLib = render(
          component({
            type: 'boxed',
            textInputProps: {
              maxLength: 10,
            },
          }),
        );

        expect(json).toMatchSnapshot();
        expect(json).toBeTruthy();
      });
    });

    describe('Type: [row]', () => {
      it('should render without crashing', () => {
        testingLib = render(component());

        const json = testingLib.toJSON();

        expect(json).toBeTruthy();

        testingLib = render(
          component({
            type: 'row',
          }),
        );

        expect(json).toMatchSnapshot();
        expect(json).toBeTruthy();
      });

      it('should render `errorText`', () => {
        testingLib = render(component());

        const json = testingLib.toJSON();

        expect(json).toBeTruthy();

        testingLib = render(
          component({
            type: 'row',
            errorText: 'error',
          }),
        );

        expect(json).toMatchSnapshot();
        expect(json).toBeTruthy();
      });

      it('should render `!editable` status', () => {
        testingLib = render(component());

        const json = testingLib.toJSON();

        expect(json).toBeTruthy();

        testingLib = render(
          component({
            type: 'row',
            editable: false,
          }),
        );

        expect(json).toMatchSnapshot();
        expect(json).toBeTruthy();
      });

      it('should render `maxLength` counter', () => {
        testingLib = render(component());

        const json = testingLib.toJSON();

        expect(json).toBeTruthy();

        testingLib = render(
          component({
            type: 'row',
            textInputProps: {
              maxLength: 10,
            },
          }),
        );

        expect(json).toMatchSnapshot();
        expect(json).toBeTruthy();
      });
    });

    describe('!editable', () => {
      it('renders without crashing', () => {
        testingLib = render(
          component({
            type: 'column',
            editable: false,
          }),
        );

        const json = testingLib.toJSON();

        expect(json).toMatchSnapshot();
        expect(json).toBeTruthy();
      });

      it('renders `labelText`', () => {
        testingLib = render(
          component({
            type: 'column',
            editable: false,
            labelText: 'label',
          }),
        );

        const json = testingLib.toJSON();

        expect(json).toMatchSnapshot();
        expect(json).toBeTruthy();
      });

      it('renders `errorText`', () => {
        testingLib = render(
          component({
            type: 'column',
            editable: false,
            labelText: 'label',
            errorText: 'error',
          }),
        );

        const json = testingLib.toJSON();

        expect(json).toMatchSnapshot();
        expect(json).toBeTruthy();
      });
    });

    //? Below tests is emitting console error but this is expeted
    describe('web', () => {
      beforeAll(() => {
        jest.mock('react-native/Libraries/Utilities/Platform', () => ({
          OS: 'web',
          select: () => ({
            web: true,
            default: undefined,
          }),
        }));
      });

      it('renders without crashing', () => {
        testingLib = render(
          component({
            type: 'column',
          }),
        );

        const json = testingLib.toJSON();

        expect(json).toMatchSnapshot();
        expect(json).toBeTruthy();
      });
    });
  });
});
