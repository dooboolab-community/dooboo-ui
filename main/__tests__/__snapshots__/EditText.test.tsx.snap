// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[EditText] interactions Type: [column] renders custom styles without crashing 1`] = `
<View
  style={
    Array [
      Object {
        "alignSelf": "stretch",
        "flexDirection": "column",
      },
      undefined,
    ]
  }
  testID="container-id"
>
  <View
    style={
      Array [
        Array [
          Object {
            "alignSelf": "stretch",
            "borderBottomColor": "#C4C4C4",
            "borderBottomWidth": 0.5,
            "flexDirection": "column",
            "justifyContent": "space-between",
          },
          Object {},
        ],
        false,
        Object {
          "borderBottomColor": "#C4C4C4",
          "borderColor": "#C4C4C4",
        },
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      editable={true}
      onBlur={[Function]}
      onFocus={[Function]}
      placeholderColor="#6D6D6D"
      rejectResponderTermination={true}
      secureTextEntry={false}
      style={
        Array [
          Array [
            Object {
              "color": "black",
              "fontSize": 14,
              "fontWeight": "bold",
              "paddingVertical": 12,
            },
            Object {},
          ],
          Object {},
          false,
        ]
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
</View>
`;

exports[`[EditText] interactions Type: [column] renders without crashing 1`] = `
<View
  style={
    Array [
      Object {
        "alignSelf": "stretch",
        "flexDirection": "column",
      },
      undefined,
    ]
  }
  testID="container-id"
>
  <View
    style={
      Array [
        Array [
          Object {
            "alignSelf": "stretch",
            "borderBottomColor": "#C4C4C4",
            "borderBottomWidth": 0.5,
            "flexDirection": "column",
            "justifyContent": "space-between",
          },
          undefined,
        ],
        false,
        Object {
          "borderBottomColor": "#C4C4C4",
          "borderColor": "#C4C4C4",
        },
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      editable={true}
      onBlur={[Function]}
      onFocus={[Function]}
      placeholderColor="#6D6D6D"
      rejectResponderTermination={true}
      secureTextEntry={false}
      style={
        Array [
          Array [
            Object {
              "color": "black",
              "fontSize": 14,
              "fontWeight": "bold",
              "paddingVertical": 12,
            },
            undefined,
          ],
          undefined,
          false,
        ]
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
</View>
`;

exports[`[EditText] interactions Type: [row] - default should render without crashing 1`] = `
<View
  style={
    Array [
      Object {
        "alignSelf": "stretch",
        "flexDirection": "column",
      },
      undefined,
    ]
  }
  testID="container-id"
>
  <View
    style={
      Array [
        Array [
          Object {
            "alignItems": "center",
            "alignSelf": "stretch",
            "borderBottomColor": "#C4C4C4",
            "borderBottomWidth": 0.5,
            "flexDirection": "row",
            "justifyContent": "space-between",
          },
          undefined,
        ],
        false,
        Object {
          "borderBottomColor": "#C4C4C4",
          "borderColor": "#C4C4C4",
        },
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      editable={true}
      onBlur={[Function]}
      onFocus={[Function]}
      placeholderColor="#6D6D6D"
      rejectResponderTermination={true}
      secureTextEntry={false}
      style={
        Array [
          Array [
            Object {
              "color": "black",
              "flex": 1,
              "fontSize": 14,
              "fontWeight": "bold",
              "paddingHorizontal": 8,
              "paddingVertical": 12,
            },
            undefined,
          ],
          undefined,
          false,
        ]
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
</View>
`;

exports[`[EditText] interactions web renders without crashing 1`] = `
<View
  style={
    Array [
      Object {
        "alignSelf": "stretch",
        "flexDirection": "column",
      },
      undefined,
    ]
  }
  testID="container-id"
>
  <View
    style={
      Array [
        Array [
          Object {
            "alignSelf": "stretch",
            "borderBottomColor": "#C4C4C4",
            "borderBottomWidth": 0.5,
            "flexDirection": "column",
            "justifyContent": "space-between",
          },
          undefined,
        ],
        false,
        Object {
          "borderBottomColor": "#C4C4C4",
          "borderColor": "#C4C4C4",
        },
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      editable={true}
      onBlur={[Function]}
      onFocus={[Function]}
      placeholderColor="#6D6D6D"
      rejectResponderTermination={true}
      secureTextEntry={false}
      style={
        Array [
          Array [
            Object {
              "color": "black",
              "fontSize": 14,
              "fontWeight": "bold",
              "paddingVertical": 12,
            },
            undefined,
          ],
          undefined,
          Object {
            "outlineWidth": 0,
          },
        ]
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
</View>
`;
